### PR TITLE
Add websocket heartbeat

### PR DESCRIPTION
**Pull request by dholms**
Add a recurring heartbeat to the websocket

Every interval (default 10s) a "ping" is sent
If a "pong" is not sent back within the interval, the websocket is terminated & restarted

Also refactored the code such that we have a `WebSocketKeepAlive` class with all of our heartbeat & retry logic and then a `Subscription` class with the smarts around xrpc.

### CODE REVIEW

File: packages/xrpc-server/src/stream/websocket-keepalive.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        this.clearHeartbeatInterval()
+      })
+
+      // send a ping on an interval and check to ensure that a pong has come back in that time
+      this.clearHeartbeatInterval() // clean up any previous interval that's stuck around
+      this.isAlive = true
+      // send a ping every x seconds
+      this.heartbeatInterval = setInterval(() => {
+        if (!this.isAlive) {
+          this.ws?.terminate()
```
Comment (devinivy) [inline packages/xrpc-server/src/stream/websocket-keepalive.ts:62]:
```suggestion
          return this.ws?.terminate()
```

File: packages/xrpc-server/src/stream/websocket-keepalive.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        } else {
+          throw err
+        }
+      }
+      break // Other side cleanly ended stream and disconnected
+    }
+  }
+
+  clearHeartbeatInterval() {
+    if (this.heartbeatInterval === null) return
```
Comment (devinivy) [inline packages/xrpc-server/src/stream/websocket-keepalive.ts:100]: `heartbeatInterval` isn't initialized to `null`, so this condition might not get triggered when it's intended to.  We could initialize `heartbeatInterval` or make this check looser, but I also don't really think the check is needed if we want to just nix it.

File: packages/xrpc-server/src/stream/websocket-keepalive.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  public reconnects: number | null = null
+  public isAlive = false
+  public heartbeatInterval: NodeJS.Timer | null
+
+  constructor(
+    public opts: ClientOptions & {
+      getUrl: () => Promise<string>
+      maxReconnectSeconds?: number
+      signal?: AbortSignal
+      heartbeatInterval?: number
```
Comment (devinivy) [inline packages/xrpc-server/src/stream/websocket-keepalive.ts:18]: Kind of funny naming conundrum here, where "interval" is sort of used in different senses in the option (a duration) versus the class property (a node timer).  Maybe just making the option `heartbeatIntervalMs` or something like that would be a way to distinguish them.

File: packages/xrpc-server/src/stream/websocket-keepalive.ts (hunk 2 above)
Comment (dholms) [reply packages/xrpc-server/src/stream/websocket-keepalive.ts:105]: ah good catch 🙏

File: packages/xrpc-server/src/stream/websocket-keepalive.ts (hunk 3 above)
Comment (dholms) [reply packages/xrpc-server/src/stream/websocket-keepalive.ts:18]: nice i like it 👍

File: packages/xrpc-server/tests/subscriptions.test.ts [hunk 4]
```diff
@@ -344,4 +345,59 @@ describe('Subscriptions', () => {
       ])
     })
   })
+
+  it.only('uses a heartbeat to reconnect if a connection is dropped', async () => {
```
Comment (devinivy) [inline packages/xrpc-server/tests/subscriptions.test.ts:349]:
```suggestion
  it('uses a heartbeat to reconnect if a connection is dropped', async () => {
```

**Comment by bnewbold** [comment]
Flagging that we should do dev comms/outreach on this if it is going to apply to the main prod firehose.

**Comment by dholms** [comment]
> Flagging that we should do dev comms/outreach on this if it is going to apply to the main prod firehose.

Thx for taking a look 🙏 

This is only on the client & is basically an improvement of the keepalive/retry logic. Doesn't affect server implementations (since ping/pong is part of the websockets spec) & not _required_ in client implementations. Just a helpful tool to detect dead connections

**Review by devinivy — APPROVED** [review]

**Pull request merged**